Common/Field.cs

```diff
@@ -24,28 +24,128 @@ namespace QuantConnect
     /// </summary>
     public static partial class Field
     {
+        private readonly static Func<IBaseData, decimal> _high = BaseDataBarPropertyOrValue(x => x.High);
+        private readonly static Func<IBaseData, decimal> _low = BaseDataBarPropertyOrValue(x => x.Low);
+        private readonly static Func<IBaseData, decimal> _open = BaseDataBarPropertyOrValue(x => x.Open);
+        private readonly static Func<IBaseData, decimal> _bidClose = BaseDataBarPropertyOrValue(x => ((QuoteBar)x).Bid.Close);
+        private readonly static Func<IBaseData, decimal> _bidOpen = BaseDataBarPropertyOrValue(x => ((QuoteBar)x).Bid.Open);
+        private readonly static Func<IBaseData, decimal> _bidLow = BaseDataBarPropertyOrValue(x => ((QuoteBar)x).Bid.Low);
+        private readonly static Func<IBaseData, decimal> _bidHigh = BaseDataBarPropertyOrValue(x => ((QuoteBar)x).Bid.High);
+        private readonly static Func<IBaseData, decimal> _askClose = BaseDataBarPropertyOrValue(x => ((QuoteBar)x).Ask.Close);
+        private readonly static Func<IBaseData, decimal> _askOpen = BaseDataBarPropertyOrValue(x => ((QuoteBar)x).Ask.Open);
+        private readonly static Func<IBaseData, decimal> _askLow = BaseDataBarPropertyOrValue(x => ((QuoteBar)x).Ask.Low);
+        private readonly static Func<IBaseData, decimal> _askHigh = BaseDataBarPropertyOrValue(x => ((QuoteBar)x).Ask.High);
+        private readonly static Func<IBaseData, decimal> _bidPrice = TickPropertyOrValue(x => ((Tick)x).BidPrice);
+        private readonly static Func<IBaseData, decimal> _askPrice = TickPropertyOrValue(x => ((Tick)x).AskPrice);
+        private readonly static Func<IBaseData, decimal> _volume = BaseDataBarPropertyOrValue(x => ((TradeBar)x).Volume);
```

Comment:
These casts will throw if not the right type and the default selector will not be used, right?

Reply:
Yup, that's right